topydo/ui/columns/CommandLineWidget.py

```diff
@@ -71,12 +95,114 @@ def _history_back(self):
         if self.history_pos != 0:
             self._history_move(-1)
 
+    def insert_completion(self, p_insert):
+        start, end = self._surrounding_text
+        final_text = start + p_insert + end
+        self.set_edit_text(final_text)
+        self.set_edit_pos(len(start) + len(p_insert))
+
+    @property
+    def completion_mode(self):
+        return len(self.completion_box) > 1
+
+    @completion_mode.setter
+    def completion_mode(self, p_enable):
+        if p_enable is True:
+            urwid.emit_signal(self, 'show_completions')
+        elif p_enable is False:
+            self._surrounding_text = None
+            if self.completion_mode:
+                self.completion_box.clear()
+                urwid.emit_signal(self, 'hide_completions')
+
+    def _complete(self):
+        pos = self.edit_pos
+        text = self.edit_text
+        completer = self.completer
+
+        word_before_cursor, is_first, start = _get_word_before_pos(text, pos)
+        completions = completer.get_completions(word_before_cursor, is_first)
+        # store slices before and after place for completion
+        self._surrounding_text = (text[:start], text[pos:])
+        single_completion = len(completions) == 1
+
+        if (single_completion == 1 and completions[0] == word_before_cursor
```

Comment:
This should work, but it would be more appropriate to rephrase it:

` if (single_completion and completions[0] == word_before_cursor or not completions`